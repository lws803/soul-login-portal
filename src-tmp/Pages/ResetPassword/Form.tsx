import { useState } from 'react';
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
  Progress,
} from '@chakra-ui/react';
import axios from 'axios';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';

import { BASE_URL } from '../../constants';

export default function Form({ token, setErrors, setIsSuccess }: Props) {
  const [isResetting, setIsResetting] = useState(false);

  const formik = useFormik({
    initialValues: {
      password: '',
    },
    onSubmit: (values) => {
      setErrors([]);
      setIsResetting(true);

      axios
        .post(`${BASE_URL}/v1/users/password-reset?token=${token}`, values)
        .then(() => setIsSuccess(true))
        .catch(({ response: { data, status } }) => {
          if (status === 400 && data.error === 'VALIDATION_ERROR') {
            formik.setErrors({
              password: 'Password is too weak!',
            });
          } else {
            setErrors([data.message]);
          }
        })
        .finally(() => {
          setIsResetting(false);
        });
    },
    validationSchema: Yup.object({
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          disabled={!token}
          aria-label="Password input"
          variant="filled"
          placeholder="Enter your new password"
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={isResetting}
          disabled={!token}
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
          width="100%"
        >
          Reset password
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (error: string[]) => void;
  setIsSuccess: (isSuccess: boolean) => void;
  token?: string | null;
};
