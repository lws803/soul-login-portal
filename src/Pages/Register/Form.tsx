import { useState } from 'react';
import {
  FormControl,
  FormLabel,
  Input,
  FormHelperText,
  FormErrorMessage,
  Box,
  Progress,
} from '@chakra-ui/react';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';
import { useFormik } from 'formik';
import { useNavigate, useLocation } from 'react-router-dom';

import FancyButton from 'src/components/FancyButton';

import { register } from './api';

export default function Form({ setErrors }: Props) {
  const navigate = useNavigate();
  const { search } = useLocation();

  const [isRegistering, setIsRegistering] = useState(false);

  const formik = useFormik({
    initialValues: {
      email: '',
      password: '',
      username: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      setIsRegistering(true);

      const { data, error } = await register({ values });
      if (error) {
        setIsRegistering(false);
        if (error.status === 400 && error.data.error === 'VALIDATION_ERROR') {
          formik.setErrors({
            password: 'Password is too weak!',
          });
        } else if (error.status === 409) {
          formik.setErrors({ email: error.data.message });
        } else {
          setErrors([error.data.message]);
        }
      }
      if (data?.id) {
        navigate(`/${search}`);
      }
    },
    validationSchema: Yup.object({
      email: Yup.string().email().required(),
      username: Yup.string().required(),
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.username && formik.touched.username}
      >
        <FormLabel htmlFor="username">Username</FormLabel>
        <Input
          id="username"
          name="username"
          onChange={formik.handleChange}
          value={formik.values.username}
          aria-label="Username input"
          variant="filled"
          placeholder="Your username"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>Choose an awesome username!</FormHelperText>
        )}
        {formik.errors.username && (
          <FormErrorMessage>{formik.errors.username}</FormErrorMessage>
        )}
      </FormControl>

      <FormControl
        isInvalid={!!formik.errors.email && formik.touched.email}
        marginTop={8}
      >
        <FormLabel htmlFor="email">Email</FormLabel>
        <Input
          id="email"
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          aria-label="Email input"
          variant="filled"
          placeholder="Your email"
          disabled={isRegistering}
        />
        {!formik.errors.email && (
          <FormHelperText>We&apos;ll never share your email.</FormHelperText>
        )}
        {formik.errors.email && (
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        )}
      </FormControl>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
        marginTop={8}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          aria-label="Password input"
          variant="filled"
          placeholder="Your password"
          disabled={isRegistering}
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <Box mt={8}>
        <FancyButton
          isLoading={isRegistering}
          disabled={isRegistering}
          type="submit"
        >
          Register
        </FancyButton>
      </Box>
    </form>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
};
