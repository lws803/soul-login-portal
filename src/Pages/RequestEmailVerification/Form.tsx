import { useFormik } from 'formik';
import * as Yup from 'yup';
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
} from '@chakra-ui/react';

import { sendEmailConfirmation } from './api';

export default function Form({ setErrors, setIsSuccess }: Props) {
  const formik = useFormik({
    initialValues: {
      email: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      const { error } = await sendEmailConfirmation(values);

      if (error) {
        setErrors([error.message]);
      } else {
        setIsSuccess(true);
      }
    },
    validationSchema: Yup.object({
      email: Yup.string().email().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.email && formik.touched.email}
        marginTop={8}
      >
        <FormLabel htmlFor="email">Email</FormLabel>
        <Input
          id="email"
          name="email"
          onChange={formik.handleChange}
          value={formik.values.email}
          aria-label="Email input"
          variant="filled"
          placeholder="Enter your email"
        />
        {formik.errors.email && (
          <FormErrorMessage>{formik.errors.email}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={formik.isSubmitting}
          loadingText="Sending verification email"
          width="100%"
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
        >
          Request email verification
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
  setIsSuccess(isSuccess: boolean): void;
};
