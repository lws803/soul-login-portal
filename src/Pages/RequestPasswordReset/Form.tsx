import { Formik, Form as FormikForm, Field, FieldProps } from 'formik';
import * as Yup from 'yup';
import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
} from '@chakra-ui/react';
import { useMutation } from '@tanstack/react-query';
import { requestPasswordReset } from 'src/modules/passwordReset';
import { ApiError } from 'src/shared/apiTypes';

export default function Form({ setErrors, setIsSuccess }: Props) {
  const { mutate: dispatchPasswordRequest } = useMutation<
    unknown,
    ApiError,
    string
  >((email) => requestPasswordReset(email), {
    onMutate: () => setErrors([]),
    onError: (error) => setErrors([error.message]),
    onSuccess: () => setIsSuccess(true),
  });

  return (
    <Formik
      initialValues={{ email: '' }}
      onSubmit={(values) => dispatchPasswordRequest(values.email)}
      validationSchema={Yup.object({
        email: Yup.string().email().required(),
      })}
      validateOnChange={false}
    >
      {({ isSubmitting }) => (
        <FormikForm>
          <Field name="email">
            {({ field, form }: FieldProps<string>) => (
              <FormControl
                isInvalid={!!form.errors.email && !!form.touched.email}
                marginTop={8}
              >
                <FormLabel htmlFor="email">Email</FormLabel>
                <Input
                  {...field}
                  aria-label="Email input"
                  variant="filled"
                  placeholder="Enter your email"
                />
                {form.errors.email && (
                  <FormErrorMessage>{form.errors.email}</FormErrorMessage>
                )}
              </FormControl>
            )}
          </Field>
          <HStack mt={8}>
            <Button
              type="submit"
              isLoading={isSubmitting}
              loadingText="Sending password reset email"
              width="100%"
              bg="soul.pink.light"
              _hover={{ bg: 'soul.pink.dark' }}
              _active={{ bg: 'soul.pink.dark' }}
            >
              Request password reset
            </Button>
          </HStack>
        </FormikForm>
      )}
    </Formik>
  );
}

type Props = {
  setErrors: (errors: string[]) => void;
  setIsSuccess(isSuccess: boolean): void;
};
