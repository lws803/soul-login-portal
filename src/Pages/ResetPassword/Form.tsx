import {
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  HStack,
  Button,
  Progress,
} from '@chakra-ui/react';
import { useFormik } from 'formik';
import * as Yup from 'yup';
import zxcvbn from 'zxcvbn';

import { resetPassword } from './api';

export default function Form({ token, setErrors, setIsSuccess }: Props) {
  const formik = useFormik({
    initialValues: {
      password: '',
    },
    onSubmit: async (values) => {
      setErrors([]);
      const { error } = await resetPassword({
        token: token!,
        newPassword: values.password,
      });

      if (!error) setIsSuccess(true);
      if (error) {
        if (error.error === 'VALIDATION_ERROR') {
          formik.setErrors({
            password: 'Password is too weak!',
          });
        } else {
          setErrors([error.message]);
        }
      }
    },
    validationSchema: Yup.object({
      password: Yup.string().required(),
    }),
    validateOnChange: false,
  });

  return (
    <form onSubmit={(e) => formik.handleSubmit(e as any)}>
      <FormControl
        isInvalid={!!formik.errors.password && formik.touched.password}
      >
        <FormLabel htmlFor="password">Password</FormLabel>
        {formik.values.password.length > 0 && (
          <Progress
            value={((zxcvbn(formik.values.password).score + 1) / 5) * 100}
            colorScheme="soul.green"
            size="xs"
            mb={1}
          />
        )}
        <Input
          id="password"
          name="password"
          type="password"
          onChange={formik.handleChange}
          value={formik.values.password}
          disabled={!token}
          aria-label="Password input"
          variant="filled"
          placeholder="Enter your new password"
        />
        {formik.errors.password && (
          <FormErrorMessage>{formik.errors.password}</FormErrorMessage>
        )}
      </FormControl>
      <HStack mt={8}>
        <Button
          type="submit"
          isLoading={formik.isSubmitting}
          disabled={!token}
          bg="soul.pink.light"
          _hover={{ bg: 'soul.pink.dark' }}
          _active={{ bg: 'soul.pink.dark' }}
          width="100%"
        >
          Reset password
        </Button>
      </HStack>
    </form>
  );
}

type Props = {
  setErrors: (error: string[]) => void;
  setIsSuccess: (isSuccess: boolean) => void;
  token?: string | null;
};
