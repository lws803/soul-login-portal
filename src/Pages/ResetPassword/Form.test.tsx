import React from 'react';
import { render, waitFor, fireEvent, act } from '@testing-library/react';
import axios from 'axios';

import Form from './Form';

describe(Form, () => {
  const props: React.ComponentProps<typeof Form> = {
    setErrors: jest.fn(),
    setIsSuccess: jest.fn(),
    token: 'TOKEN',
  };

  beforeEach(() => {
    jest
      .spyOn(axios, 'post')
      .mockResolvedValue({ data: { error: null, status: 200 } });
  });

  it('renders', async () => {
    const { container } = render(<Form {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('resets password', async () => {
    const { getByLabelText, getByText } = render(<Form {...props} />);

    act(() => {
      fireEvent.change(getByLabelText('Password input'), {
        target: { value: 'NEW_PASSWORD' },
      });
    });

    fireEvent.click(getByText('Reset password'));

    await waitFor(() => {
      expect(axios.post).toHaveBeenCalledWith(
        'http://api.network.com/v1/users/password-reset?token=TOKEN',
        { password: 'NEW_PASSWORD' },
      );
    });
  });
});
