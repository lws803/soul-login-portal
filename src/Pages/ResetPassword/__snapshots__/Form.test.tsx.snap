// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form renders 1`] = `
<div>
  <form>
    <div
      class="chakra-form-control css-0"
      role="group"
    >
      <label
        class="chakra-form__label css-uvho5g"
        for="password"
        id="field-:r0:-label"
      >
        Password
      </label>
      <input
        aria-label="Password input"
        class="chakra-input css-0"
        id="password"
        name="password"
        placeholder="Enter your new password"
        type="password"
        value=""
      />
    </div>
    <div
      class="chakra-stack css-nzla2i"
    >
      <button
        class="chakra-button css-1k6167g"
        type="submit"
      >
        Reset password
      </button>
    </div>
  </form>
</div>
`;
